fabric-dimensions-v1/src/main/java/net/fabricmc/fabric/impl/dimension/DimensionIdsFixer.java

```diff
@@ -0,0 +1,119 @@
+package net.fabricmc.fabric.impl.dimension;
+
+import io.netty.buffer.Unpooled;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import net.fabricmc.fabric.api.dimension.FabricDimensionType;
+import net.fabricmc.fabric.api.network.PacketContext;
+import net.fabricmc.fabric.api.network.ServerSidePacketRegistry;
+import net.fabricmc.fabric.impl.registry.RemapException;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.network.Packet;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.PacketByteBuf;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.world.dimension.DimensionType;
+import net.minecraft.world.level.LevelProperties;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Consumer;
+
+/**
+ * Handles fixing raw dimension ids between saves and servers
+ */
+public class DimensionIdsFixer {
+    static final Identifier ID = new Identifier("fabric", "dimension/sync");
+
+    /**
+     * Assigns a unique id to every registered {@link FabricDimensionType}, keeping the known ids
+     * from {@code savedIds}.
+     * @param savedIds a compound tag mapping dimension ids to raw ids
+     * @return id to raw id mappings of the current instance
+     * @throws RemapException if dimensions IDs conflict irredeemably
+     */
+    public static CompoundTag apply(CompoundTag savedIds) throws RemapException {
```

Comment:
The method could be more readable if it didn't use id to refer to both the int id ("rawId") and the string id/identifier/name ("[dim]id").